# Question 1
<p style="font-size:22px;">
Run K-means on the datasets. And record its SSE
</p>

<p style="font-size:20px;">
First, read the data and delete the Norminal column, and convert the data to a numpy array
</p>

In [1]:
import pandas as pd
import numpy as np


data = pd.read_csv('clustering_data.csv')

# delete the first column and convert to numpy array
price_changes = data.iloc[:, 1:].values

price_changes

array([[ -4.7557   ,   4.00509  ,   3.58155  ,  -0.395257 ,   0.768624 ,
          1.12594  ,  -0.237274 ,  -1.91728  ,   0.706794 ,   0.957592 ,
         -0.0206825,   6.34668  ,  -2.82923  ,   1.68118  ,  -4.10586  ,
          4.63801  ,   1.46898  ,   2.74809  ,  -0.0228676,   1.87709  ,
         -0.70247  ,   2.44804  ,  -3.13752  ,  -1.13863  ,  -0.0651749],
       [ -3.2019   ,   5.65488  ,  -1.44928  ,   0.693878 ,   0.574788 ,
          1.50561  ,  -1.42566  ,   0.467675 ,  -2.90853  ,   0.0558659,
         -3.65062  ,   0.99398  ,  -2.76886  ,  -1.29434  ,  -3.80463  ,
          0.93633  ,   0.122649 ,   3.74037  ,  -0.92452  ,   4.33917  ,
          3.06093  ,   4.88284  ,  -0.0691085,  -0.353045 ,   1.15721  ],
       [ -0.55384  ,   1.92791  ,   0.529256 ,   1.80451  ,   2.05676  ,
         -0.869652 ,  -3.18936  ,   3.37173  ,   3.67084  ,  -4.0242   ,
         -0.9911   ,  -0.853207 ,  -1.21903  ,  -6.0285   ,  -3.45091  ,
          2.06707  ,   1.0505   ,   3.03001  ,   

In [27]:
from sklearn.cluster import KMeans

kmeans = KMeans(init='random', n_clusters=8)
kmeans.fit(price_changes)


C:\Users\19928\.conda\envs\dm\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(init='random')

<p style="font-size:20px;">
Shows for each data points its cluster.
</p>

In [51]:
kmeans.labels_, kmeans.cluster_centers_, kmeans.n_iter_


(array([1, 1, 0, 3, 4, 5, 4, 2, 5, 1, 1, 1, 1, 5, 4, 7, 6, 1, 1, 5, 5, 0,
        1, 0, 1, 1, 5, 5, 1, 1], dtype=int32),
 array([[ -0.45276167,   2.99973333,   0.497452  ,   0.862148  ,
           2.40279667,  -0.39941267,  -1.73934787,   2.7955    ,
           2.14000333,  -3.07447333,  -1.13949867,  -1.29716333,
          -2.01673   ,  -4.98672333,  -2.24480833,   2.11092233,
           0.872748  ,   1.44407333,   1.42131167,   1.698886  ,
           4.59336   ,   1.879895  ,   1.07392867,  -0.50375667,
           0.82935497],
        [ -0.68975623,   2.50959692,  -0.30901777,   0.63056008,
           0.64237339,   0.50161768,  -0.4863914 ,   0.90291337,
          -1.83279615,  -0.56041832,  -0.72864611,   2.24029046,
          -1.50509485,  -1.03314131,  -3.52789692,   2.14831131,
          -0.02605021,   2.580847  ,  -0.53053835,   1.89575738,
           1.21996823,   0.5509682 ,  -1.75866081,  -0.643416  ,
           1.35087855],
        [ -4.5614    ,  -0.324675  ,  -2.60723   , 

In [28]:
# SSE
sse = kmeans.inertia_
print("Sum of Squared Errors (SSE):", sse)

Sum of Squared Errors (SSE): 1685.7398582935516


<p style="font-size:20px;">
So here, We only run a most simply K-means method with init='random' and K=8. And We get the centroids (centers) and the cluster of each data points.<br>
And also, We get the SSE=1685.7398582935516
</p>

In [57]:
from sklearn.metrics import silhouette_score

# Silhouette Score
labels = kmeans.labels_
silhouette_avg = silhouette_score(price_changes, labels)
print(f'Silhouette Score: {silhouette_avg:.3f}')


Silhouette Score: 0.095


<p style="font-size:20px;">
Through the Silhouette Score evaluation, we found that the most simple K-means clustering effect is not so good.
</p>

# Question 2
<p style="font-size:22px;">
Change the parameters and decrease the SSE as much as possible.
</p>

<p style="font-size:20px;">
Due to the wording of the question, it is recommended that we choose K=8, which means there are a total of 8 clusters. Therefore, we can modify a total of four parameters, namely:<br><br>
The K-means method (init), <br>
Maximum iteration count (max_iter), <br>
Initialization repetition count (n_init), <br>
Convergence tolerance (tol)<br>
</p>

In [62]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans
from itertools import product


# set the range of the method and the parameters
init_methods = ['random', 'k-means++']
n_init_values = [10, 50, 100]
max_iter_values = [2, 3, 4]
tol_values = [1e-4, 1e-16]

results = []


# generate all parameters group
for init, n_init, max_iter, tol in product(init_methods, n_init_values, max_iter_values, tol_values):
    
    if max_iter in max_iter_values or tol in tol_values:
        
        kmeans = KMeans(
            n_clusters=8,
            init=init,
            n_init=n_init,
            max_iter=max_iter,
            tol=tol,
            random_state=520
        )
    else:
        kmeans = KMeans(
            n_clusters=8,
            init=init,
            n_init=n_init,
            max_iter=max_iter,
            tol=tol
        )
    kmeans.fit(price_changes)
    
    sse = kmeans.inertia_
    
    results.append({
        'init': init,
        'n_init': n_init,
        'max_iter': max_iter,
        'tol': tol,
        'SSE': sse
    })


results_df = pd.DataFrame(results)

# sort and index and show
results_df.sort_values(by='SSE', ascending=True, inplace=True)
results_df.reset_index(drop=True, inplace=True)

results_df


C:\Users\19928\.conda\envs\dm\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\19928\.conda\envs\dm\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\19928\.conda\envs\dm\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\19928\.conda\envs\dm\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a m

init  n_init  max_iter           tol          SSE
0   k-means++     100         4  1.000000e-16  1527.455307
1   k-means++     100         4  1.000000e-04  1527.455307
2   k-means++     100         3  1.000000e-16  1527.455307
3   k-means++     100         3  1.000000e-04  1527.455307
4   k-means++      50         2  1.000000e-04  1537.246760
5   k-means++      50         2  1.000000e-16  1537.246760
6   k-means++      50         3  1.000000e-04  1537.246760
7   k-means++      50         3  1.000000e-16  1537.246760
8   k-means++      50         4  1.000000e-04  1537.246760
9   k-means++      50         4  1.000000e-16  1537.246760
10  k-means++     100         2  1.000000e-04  1537.246760
11  k-means++     100         2  1.000000e-16  1537.246760
12  k-means++      10         4  1.000000e-16  1542.844075
13  k-means++      10         4  1.000000e-04  1542.844075
14  k-means++      10         3  1.000000e-16  1542.844075
15  k-means++      10         3  1.000000e-04  1542.844075
16  k-means++      10         2  1.000000e-04  1555.632869
17  k-means++      10         2  1.000000e-16  1555.632869
18     random      50         3  1.000000e-04  1620.131890
19     random      50         3  1.000000e-16  1620.131890
20     random     100         3  1.000000e-16  1620.131890
21     random     100         3  1.000000e-04  1620.131890
22     random      50         4  1.000000e-16  1620.131890
23     random      50         4  1.000000e-04  1620.131890
24     random     100         4  1.000000e-16  1620.131890
25     random     100         4  1.000000e-04  1620.131890
26     random      50         2  1.000000e-16  1643.294207
27     random      50         2  1.000000e-04  1643.294207
28     random     100         2  1.000000e-04  1643.294207
29     random     100         2  1.000000e-16  1643.294207
30     random      10         3  1.000000e-16  1716.867150
31     random      10         3  1.000000e-04  1716.867150
32     random      10         4  1.000000e-04  1716.867150
33     random      10         4  1.000000e-16  1716.867150
34     random      10         2  1.000000e-16  1728.472629
35     random      10         2  1.000000e-04  1728.472629

<p style="font-size:20px;">
Here, we find that, the K-means++ method always better than the initial K-means method. However, this does not mean that K-means++will always be better than the K-means method. Because the K-means++method only increases the probability of selecting points with initial distance as centroids, it can achieve better performance than ordinary K-means methods in most cases. However, there is still a situation where the K-means method selects the most suitable centroid at a certain time, which the K-means++method does not have.<br><br>
And then, we can mining more infomation of the trend.
</p>

<p style="font-size:20px;">
Here, we set three parameters fixed, and the last remaining parameter is variable, so we can see how SSE changes with the variation of different parameters.
</p>

In [63]:
from itertools import combinations

def view_all_sse_by_varying_parameters(results_df):

    fixed_init = 'k-means++'
    
    param_names = ['n_init', 'max_iter', 'tol']
    
    all_results = {}
    
    # 3 fixed, 1 changed
    for fixed_params in combinations(param_names, 2):
        
        # changed parameters
        varying_param = [param for param in param_names if param not in fixed_params][0]
        
        # fixed parameters group
        fixed_param_combinations = results_df[(results_df['init'] == fixed_init)][list(fixed_params)].drop_duplicates()
        
        # all groups
        for _, fixed_values in fixed_param_combinations.iterrows():

            fixed_params_dict = dict(zip(fixed_params, fixed_values))
            fixed_params_dict['init'] = fixed_init
            
            filtered_results = results_df.copy()
            for param, value in fixed_params_dict.items():
                filtered_results = filtered_results[filtered_results[param] == value]
            
            filtered_results = filtered_results[[varying_param, 'SSE']].sort_values(by=varying_param)
            filtered_results.reset_index(drop=True, inplace=True)
            
            key = f"fixed parameters: {fixed_params_dict}, changed parameters: {varying_param}"
            all_results[key] = filtered_results
    
    return all_results

all_sse_results = view_all_sse_by_varying_parameters(results_df)

for key, df in all_sse_results.items():
    print(f"\n{key}\n")
    print(df)



fixed parameters: {'n_init': 100, 'max_iter': 4, 'init': 'k-means++'}, changed parameters: tol

            tol          SSE
0  1.000000e-16  1527.455307
1  1.000000e-04  1527.455307

fixed parameters: {'n_init': 100, 'max_iter': 3, 'init': 'k-means++'}, changed parameters: tol

            tol          SSE
0  1.000000e-16  1527.455307
1  1.000000e-04  1527.455307

fixed parameters: {'n_init': 50, 'max_iter': 2, 'init': 'k-means++'}, changed parameters: tol

            tol         SSE
0  1.000000e-16  1537.24676
1  1.000000e-04  1537.24676

fixed parameters: {'n_init': 50, 'max_iter': 3, 'init': 'k-means++'}, changed parameters: tol

            tol         SSE
0  1.000000e-16  1537.24676
1  1.000000e-04  1537.24676

fixed parameters: {'n_init': 50, 'max_iter': 4, 'init': 'k-means++'}, changed parameters: tol

            tol         SSE
0  1.000000e-16  1537.24676
1  1.000000e-04  1537.24676

fixed parameters: {'n_init': 100, 'max_iter': 2, 'init': 'k-means++'}, changed parameters: 

<p style="font-size:20px;">
From the results, it can be seen that:<br><br>
 As the maximum number of iterations (max_iter) increases, SSE decreases and stops at the end of the algorithm. That is to say, when the algorithm is not converging, it will decrease with the increase of max_iter, because the increase in the maximum number of iterations means a more suitable calculation of the centroid when the algorithm is not converging. But once the algorithm has converged within max_iter, SSE will not increase with the increase of max_iter, because the algorithm has already converged within the maximum number of iterations and has not reached the maximum number of iterations, that is, max_iter will not affect the number of iterations of the algorithm.<br><br>
 As the number of initialization iterations (n_init) increases, SSE may decrease. Because the K-means algorithm selects the cluster with the best value, i.e., the cluster with the smallest SSE, among the initialization steps of n2, as the result output, an increase in n2 implies a greater likelihood of finding a better cluster. However, this does not necessarily mean that an increase in n_inite will lead to a decrease in SSE, as the algorithm may not be able to find a better clustering solution within a larger number of initialization attempts. This can result in SSE not increasing as n_inite increases.<br><br>
 Here, SSE does not decrease with the decrease of the minimum threshold (tol) for centroid change. This may be because within the iteration times, the centroids of the 8 clusters are no longer changing, or the changes are very, very small. This will result in changes in tol not affecting the iteration times of the algorithm, nor will it affect the value of SSE.However, under normal circumstances, smaller convergence conditions lead to more iterations, resulting in smaller SSE.
</p>

# Question 3
<p style="font-size:22px;">
try to label each cluster with a topic.
</p>

In [68]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=8, init='k-means++', n_init=10000, max_iter=100)
kmeans.fit(price_changes)

# SSE
sse = kmeans.inertia_
print("Sum of Squared Errors (SSE):", sse)

C:\Users\19928\.conda\envs\dm\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


Sum of Squared Errors (SSE): 1508.434119398986


In [84]:
import pandas as pd

labels = kmeans.labels_

stock_clusters = pd.DataFrame({'Stock': data['StockName'], 'Cluster': labels})

for cluster in range(kmeans.n_clusters):
    print(f"\nCluster {cluster}:")
    stocks_in_cluster = stock_clusters[stock_clusters['Cluster'] == cluster]['Stock'].values
    print(stocks_in_cluster)



Cluster 0:
['IBM' 'The Home Depot' 'Intel' 'Wal-Mart' 'General Electric'
 'United Technologies' 'Travelers' '3M']

Cluster 1:
['American Express' 'Boeing' 'Microsoft' 'Walt Disney' 'JPMorgan Chase']

Cluster 2:
['Bank of America']

Cluster 3:
['Kraft' 'Verizon' 'Procter & Gamble' 'AT&T' 'Merck' 'McDonalds'
 'Coca-Cola' 'Johnson & Johnson']

Cluster 4:
['Cisco Systems']

Cluster 5:
['Hewlett-Packard']

Cluster 6:
['Chevron' 'Pfizer' 'ExxonMobil']

Cluster 7:
['DuPont' 'Caterpillar' 'Alcoa']



### Cluster 1:
<p style="font-size:16px;">
Contains: American Express, Boeing, Microsoft, Walt Disney, and JPMorgan Chase.
This cluster includes companies from finance, industrials, information technology, and communication services, representing major global companies.<br><br>
 So this cluster can be labeled as the “Large Multinational Corporation Stocks.”<br><br>
</p>

### Cluster 2:
<p style="font-size:16px;">
Contains: Bank of America.<br><br>
So this cluster can be labeled as the “Financial Stocks.”<br><br>
</p>

### Cluster 3:
<p style="font-size:16px;">
Contains: Kraft, Verizon, Procter & Gamble, AT&T, Merck, McDonald’s, Coca-Cola, and Johnson & Johnson.
These companies span consumer staples, communication services, healthcare, and discretionary consumption, and most of which are well-known consumer brands.<br><br>
So this cluster can be labeled as the “Consumer Stocks.”<br><br>
</p>

### Cluster 4:
<p style="font-size:16px;">
Contains: Cisco Systems.<br><br>
So this cluster can be labeled as the “Information Technology Stocks.”<br><br>
</p>

### Cluster 5:
<p style="font-size:16px;">
Contains: Hewlett-Packard.<br><br>
So this cluster can also be labeled as the “Information Technology Stocks.”<br><br>
</p>

### Cluster 6:
<p style="font-size:16px;">
Contains: Chevron, Pfizer, and ExxonMobil.
This cluster consists of companies in the energy and healthcare sectors and more in Energy.<br><br>
So this cluster can be labeled as the “Energy Stocks.”<br><br>
</p>

### Cluster 7:
<p style="font-size:16px;">
Contains: DuPont, Caterpillar, and Alcoa.
These stocks belong primarily to the materials and industrials sectors and more in Materials.<br><br>
So this cluster can be labeled as the “Materials Stocks.”<br><br>
</p>
